### ISSUE TITLE
Confusion about bucketing and steps

### ISSUE THREAD

**Issue by Oruli** [tdrussell/diffusion-pipe#401]
I am training Qwen on a dataset of 54 images, below is the output from the bucketing process (I think), and it looks as though all image are placed into a bucket. Some buckets are dropped later (quite a few) due to only having a single image.

The confusion is when it gets going, each epoch is only 12 steps, but shouldn't it be 54?

```
caching latents: (0.63, 1)
Adding size bucket: 40 examples [00:00, 6577.24 examples/s]                                                                               
caching latents: (0.63, 800, 1280, 1)
Building iteration order
Saving the dataset (1/1 shards): 100%|██████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 14680.80 examples/s]
caching latents: (1.0, 1)
Adding size bucket: 2 examples [00:00, 532.07 examples/s]                                                                                 
caching latents: (1.0, 1024, 1024, 1)
num_proc must be <= 1. Reducing num_proc to 1 for dataset of size 1.
Building iteration order
Saving the dataset (1/1 shards): 100%|██████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 836.69 examples/s]
caching latents: (1.587, 1)
Adding size bucket: 10 examples [00:00, 2590.68 examples/s]                                                                               
caching latents: (1.587, 1280, 800, 1)
num_proc must be <= 5. Reducing num_proc to 5 for dataset of size 5.
Building iteration order
Saving the dataset (1/1 shards): 100%|█████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 3555.10 examples/s]
caching latents: (0.794, 1)
Adding size bucket: 48 examples [00:00, 6909.42 examples/s]                                                                               
caching latents: (0.794, 928, 1152, 1)
Building iteration order
Saving the dataset (1/1 shards): 100%|██████████████████████████████████████████████████████████| 24/24 [00:00<00:00, 19839.04 examples/s]
caching latents: (1.26, 1)
Adding size bucket: 6 examples [00:00, 1639.04 examples/s]                                                                                
caching latents: (1.26, 1152, 928, 1)
num_proc must be <= 3. Reducing num_proc to 3 for dataset of size 3.
Building iteration order
Saving the dataset (1/1 shards): 100%|█████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 2590.68 examples/s]
caching latents: (0.5, 1)
Adding size bucket: 2 examples [00:00, 572.60 examples/s]                                                                                 
caching latents: (0.5, 736, 1440, 1)
num_proc must be <= 1. Reducing num_proc to 1 for dataset of size 1.
Building iteration order
Saving the dataset (1/1 shards): 100%|██████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 868.03 examples/s]
```

I think increased micro batch to 2, and grad_accum to 8 and now I only get 2 steps. IS this a good thing?

```
steps: 1 loss: 0.0762 iter time (s): 25.868 samples/sec: 0.619
[2025-08-28 09:08:03,005] [INFO] [logging.py:107:log_dist] [Rank 0] step=2, skipped=0, lr=[1e-06], mom=[0.0]
steps: 2 loss: 0.0967 iter time (s): 27.244 samples/sec: 0.587
Started new epoch: 2
```

**Comment by jakehemmerle** [comment]
micro batch at 2 with grad accum 8 means effective batch is 16 which is way too high for a dataset of 54. 2 steps means 32 images per epoch i think (and that makes sense since 25.868s * 0.619 = ~16) so you're losing a lot of images. Your effective batch size should be no more than 2 or so, so set micro batch at 1-2 and turn grad accum to 1.

I'd maybe try reducing the number of buckets to 3 unless you're using micro batch of 1.

If you're training a LoRA, you will prob want to use a higher learning rate like 5e-5.

**Comment by Oruli** [comment]
> micro batch at 2 with grad accum 8 means effective batch is 16 which is way too high for a dataset of 54. 2 steps means 32 images per epoch i think (and that makes sense since 25.868s * 0.619 = ~16) so you're losing a lot of images. Your effective batch size should be no more than 2 or so, so set micro batch at 1-2 and turn grad accum to 1.
> 
> I'd maybe try reducing the number of buckets to 3 unless you're using micro batch of 1.
> 
> If you're training a LoRA, you will prob want to use a higher learning rate like 5e-5.

I'm using the default example for Qwen, I wonder why grad_accum was changed to 8 there? What would be the best settings for 100 images, for 2 x 5090 GPUs? I'm currently only using 15GB VRAM on each GPU, and my understanding was you can increase batch size if you have spare VRAM, why wouldn't you?

I don't quite understand how I'm "losing a lot of images", would this be down to the bucket sizes dropping a lot or do you mean something else?

For learning rate, again I was using the current default example:  type = 'automagic'

**Comment by Oruli** [comment]
I'm getting huge fluctuations in loss, here is my config:

```
# Change these paths
output_dir = '/training/diffusion-pipe/output/set/'
dataset = '/training/diffusion-pipe/run/set.toml'

# You can have separate eval datasets. Give them a name for Tensorboard metrics.
eval_datasets = [
    {name = 'eval', config = '/training/diffusion-pipe/run/set.toml'},
]

# training settings
epochs = 1000
micro_batch_size_per_gpu = 2
pipeline_stages = 2
gradient_accumulation_steps = 1
gradient_clipping = 1
warmup_steps = 100

# eval settings
eval_every_n_epochs = 1
#eval_every_n_steps = 100
eval_before_first_step = true
eval_micro_batch_size_per_gpu = 2
eval_gradient_accumulation_steps = 1

# misc settings
save_every_n_epochs = 10
#save_every_n_steps = 250
#checkpoint_every_n_epochs = 1
checkpoint_every_n_minutes = 60
activation_checkpointing = true
partition_method = 'parameters'
save_dtype = 'bfloat16'
caching_batch_size = 8
steps_per_print = 1
#blocks_to_swap = 0

[model]
type = 'qwen_image'
# Change this path
diffusers_path = '/training_models/Qwen-Image/'
dtype = 'bfloat16'
transformer_dtype = 'float8'
timestep_sample_method = 'logit_normal'

[adapter]
type = 'lora'
rank = 16
dtype = 'bfloat16'

# Automagic should work fine and requires no tuning, but you can probably get better results using AdamW
# with a properly tuned learning rate.
[optimizer]
#type = 'automagic'
#weight_decay = 0.01

# [optimizer]
type = 'AdamW8bitKahan'
lr = 2e-5
betas = [0.9, 0.99]
weight_decay = 0.01
```

```
[2025-08-31 12:59:40,036] [INFO] [logging.py:107:log_dist] [Rank 0] step=81, skipped=0, lr=[1.6238e-05], mom=[[0.9, 0.99]]
steps: 81 loss: 0.0749 iter time (s): 5.564 samples/sec: 0.359
[2025-08-31 12:59:45,763] [INFO] [logging.py:107:log_dist] [Rank 0] step=82, skipped=0, lr=[1.6435999999999996e-05], mom=[[0.9, 0.99]]
steps: 82 loss: 0.0625 iter time (s): 5.707 samples/sec: 0.350
[2025-08-31 12:59:51,557] [INFO] [logging.py:107:log_dist] [Rank 0] step=83, skipped=0, lr=[1.6633999999999997e-05], mom=[[0.9, 0.99]]
steps: 83 loss: 0.0819 iter time (s): 5.733 samples/sec: 0.349
[2025-08-31 12:59:57,196] [INFO] [logging.py:107:log_dist] [Rank 0] step=84, skipped=0, lr=[1.6831999999999997e-05], mom=[[0.9, 0.99]]
steps: 84 loss: 0.1620 iter time (s): 5.610 samples/sec: 0.357
[2025-08-31 13:00:02,764] [INFO] [logging.py:107:log_dist] [Rank 0] step=85, skipped=0, lr=[1.7029999999999998e-05], mom=[[0.9, 0.99]]
steps: 85 loss: 0.0819 iter time (s): 5.539 samples/sec: 0.361
[2025-08-31 13:00:08,380] [INFO] [logging.py:107:log_dist] [Rank 0] step=86, skipped=0, lr=[1.7228e-05], mom=[[0.9, 0.99]]
steps: 86 loss: 0.1355 iter time (s): 5.584 samples/sec: 0.358
[2025-08-31 13:00:14,112] [INFO] [logging.py:107:log_dist] [Rank 0] step=87, skipped=0, lr=[1.7426e-05], mom=[[0.9, 0.99]]
steps: 87 loss: 0.0637 iter time (s): 5.705 samples/sec: 0.351
[2025-08-31 13:00:19,658] [INFO] [logging.py:107:log_dist] [Rank 0] step=88, skipped=0, lr=[1.7623999999999996e-05], mom=[[0.9, 0.99]]
steps: 88 loss: 0.0719 iter time (s): 5.524 samples/sec: 0.362
[2025-08-31 13:00:25,252] [INFO] [logging.py:107:log_dist] [Rank 0] step=89, skipped=0, lr=[1.7821999999999996e-05], mom=[[0.9, 0.99]]
steps: 89 loss: 0.2330 iter time (s): 5.568 samples/sec: 0.359
```

**Comment by jakehemmerle** [comment]
loss looks normal to me

**Comment by Oruli** [comment]
> loss looks normal to me

Even the fact it's jumping around so much? I did add an eval set (first time i've used that), and here is the output.

It just feels so much slower than when I wasn't using multigpu, but maybe thats the model training slower.

<img width="849" height="333" alt="Image" src="https://github.com/user-attachments/assets/e5de9813-eb8f-4f3d-9418-22c3dfb0278c" />


<img width="408" height="309" alt="Image" src="https://github.com/user-attachments/assets/6c3c09a0-e854-4698-be81-d1381c3cb161" />
